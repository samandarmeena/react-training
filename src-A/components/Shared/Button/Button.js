import React from 'react';
import PropTypes from 'prop-types';
import styles from './Button.module.scss';

/**
 * Button component
 * It renders button
 * @param {string} buttonType
 * @param {string} buttonName
 * @param {string} buttonId
 * @param {string} buttonTitle
 * @param {string} buttonStyle
 * @param {string} buttonSize
 * @param {string} addOnClass
 * @param {string} buttonToolTip
 * @param {fn} clickEvent
 * @return Element
 */
const Button = props => {
  const {
    buttonType,
    buttonName,
    buttonId,
    buttonTitle,
    buttonStyle,
    buttonSize,
    addOnClass,
    buttonToolTip,
    clickEvent
  } = props;
  return (
    <button
      type={buttonType}
      name={buttonName}
      id={buttonId}
      className={`${styles.button} ${styles[buttonStyle]} ${styles[buttonSize]} ${addOnClass}`}
      onClick={() => clickEvent()}
      title={buttonToolTip}
    >
      {buttonTitle}
    </button>
  );
};

// proptype validations for props
Button.propTypes = {
  buttonType: PropTypes.string,
  buttonName: PropTypes.string.isRequired,
  buttonId: PropTypes.string.isRequired,
  buttonTitle: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
    .isRequired,
  buttonStyle: PropTypes.string,
  buttonSize: PropTypes.string,
  addOnClass: PropTypes.string,
  buttonToolTip: PropTypes.string.isRequired,
  clickEvent: PropTypes.func
};

// default values for props
Button.defaultProps = {
  buttonType: 'button',
  buttonStyle: 'buttonPrimary',
  buttonSize: 'buttonMedium',
  addOnClass: ''
};

export default Button;
