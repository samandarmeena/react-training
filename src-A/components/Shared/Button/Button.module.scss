.button {
  border: none;
  cursor: pointer;
  margin-left: 3px;
  margin-right: 3px;
}
.buttonInfo {
  background-color: #138496;
  color: #fff;
}
.buttonInfo:hover {
  color: #fff;
  background-color: #007183;
  border-color: #007183;
}
.buttonPrimary {
  background-color: #0069d9;
  color: #fff;
}
.buttonPrimary:hover {
  background-color: #0052c2;
}
.buttonSuccess {
  background-color: #40c2a6;
  color: #fff;
}
.buttonWarning {
  background-color: #e0a800;
  color: #fff;
}
.buttonDanger {
  background-color: #c82333;
  color: #fff;
}
.buttonDanger:hover {
  background-color: #ab0616;
}
.buttonLarge {
  font-size: 17px;
  border-radius: 5px;
  padding: 13px;
}
.buttonMedium {
  font-size: 15px;
  border-radius: 5px;
  padding: 10px;
}
.buttonSmall {
  font-size: 12px;
  border-radius: 5px;
  padding: 5px;
}
