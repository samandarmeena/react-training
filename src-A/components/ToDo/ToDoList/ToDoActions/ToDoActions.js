import React from 'react';
import PropTypes from 'prop-types';
import Image from '../../../Shared/Image';
import Button from '../../../Shared/Button';

/**
 * ToDoActions component
 * renders ToDoActions container
 * @param {string} itemId
 * @param {string} itemStatus
 * @param {func} handleToggleFixed
 * @param {func} handleRemoveToDo
 * @return Element
 */
const ToDoActions = (props) => {
  const { itemId, itemStatus, handleToggleFixed, handleRemoveToDo } = props;
  const actionButtons = [
    {
      buttonName: `complete${itemId}`,
      buttonId: `complete${itemId}`,
      buttonSize: 'buttonSmall',
      buttonToolTip: itemStatus === 'New' ? 'Mark as Completed' : 'Mark as New',
      buttonStyle: itemStatus === 'New' ? 'buttonSuccess' : 'buttonInfo',
      addOnClass: 'changeStatus',
      buttonTitle:
        itemStatus === 'New' ? (
          <Image src="/assets/images/check.png" alt="Check" />
        ) : (
          <Image src="/assets/images/return.png" alt="Return" />
        ),
      clickEvent: () => handleToggleFixed(itemId)
    },
    {
      buttonName: `remove${itemId}`,
      buttonId: `remove${itemId}`,
      buttonSize: 'buttonSmall',
      buttonStyle: 'buttonDanger',
      addOnClass: 'removeToDo',
      buttonToolTip: 'Remove ToDo',
      buttonTitle: <Image src="/assets/images/trash.png" alt="Remove" />,
      clickEvent: () => handleRemoveToDo(itemId)
    }
  ];
  return (
    <>
      {actionButtons.map((actionButton) => {
        return (
          <Button
            key={actionButton.buttonId}
            buttonName={actionButton.buttonName}
            buttonId={actionButton.buttonId}
            buttonSize={actionButton.buttonSize}
            buttonStyle={actionButton.buttonStyle}
            addOnClass={actionButton.addOnClass}
            buttonToolTip={actionButton.buttonToolTip}
            buttonTitle={actionButton.buttonTitle}
            clickEvent={actionButton.clickEvent}
          />
        );
      })}
    </>
  );
};

// proptype validations for props
ToDoActions.propTypes = {
  itemId: PropTypes.number.isRequired,
  itemStatus: PropTypes.string.isRequired,
  handleToggleFixed: PropTypes.func.isRequired,
  handleRemoveToDo: PropTypes.func.isRequired
};

export default ToDoActions;
