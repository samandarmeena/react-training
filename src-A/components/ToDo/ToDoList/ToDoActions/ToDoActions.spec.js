import React from 'react';
import { mount } from 'enzyme';
import ToDoActions from './ToDoActions';
import Button from '../../../Shared/Button';

describe('ToDoActions component', () => {
  let wrapper = '';
  let inputProps = {};
  const mockToggleFixed = jest.fn();
  const mockRemoveToDo = jest.fn();
  beforeAll(() => {
    inputProps = {
      itemId: 123,
      itemStatus: 'New',
      handleToggleFixed: mockToggleFixed,
      handleRemoveToDo: mockRemoveToDo
    };
    wrapper = mount(<ToDoActions {...inputProps} />);
  });

  it('ToDoActions component rendered without an error', () => {
    expect(wrapper).toBeTruthy();
  });

  it('Check length of the children', () => {
    expect(wrapper.find(Button)).toHaveLength(2);
  });

  it('Check attributes of the children components', () => {
    expect(
      wrapper
        .find(Button)
        .find('button')
        .first()
        .hasClass('changeStatus')
    ).toBe(true);
    expect(
      wrapper
        .find(Button)
        .find('button')
        .at(1)
        .hasClass('removeToDo')
    ).toBe(true);
  });

  it('Check dynamic props of the components', () => {
    let newinputProps = {
      ...inputProps,
      itemStatus: 'new1'
    };
    wrapper = mount(<ToDoActions {...newinputProps} />);
    expect(
      wrapper
        .find(Button)
        .first()
        .prop('buttonToolTip')
    ).toBe('Mark as New');
    expect(
      wrapper
        .find(Button)
        .first()
        .prop('buttonStyle')
    ).toBe('buttonInfo');
  });

  it('mock function (toggleFix) should trigger on click', () => {
    wrapper
      .find(Button)
      .first()
      .props()
      .clickEvent();
    expect(mockToggleFixed).toHaveBeenCalled();
  });

  it('mock function (removeToDo) should trigger on click', () => {
    wrapper
      .find(Button)
      .at(1)
      .props()
      .clickEvent();
    expect(mockRemoveToDo).toHaveBeenCalled();
  });

  it('Snapshot testing', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
