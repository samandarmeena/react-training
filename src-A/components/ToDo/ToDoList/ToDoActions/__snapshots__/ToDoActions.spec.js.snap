// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoActions component Snapshot testing 1`] = `
<ToDoActions
  handleRemoveToDo={
    [MockFunction] {
      "calls": Array [
        Array [
          123,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  handleToggleFixed={
    [MockFunction] {
      "calls": Array [
        Array [
          123,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  itemId={123}
  itemStatus="new1"
>
  <Button
    addOnClass="changeStatus"
    buttonId="complete123"
    buttonName="complete123"
    buttonSize="buttonSmall"
    buttonStyle="buttonInfo"
    buttonTitle={
      <Image
        alt="Return"
        src="/assets/images/return.png"
      />
    }
    buttonToolTip="Mark as New"
    buttonType="button"
    clickEvent={[Function]}
    key="complete123"
  >
    <button
      className="button buttonInfo buttonSmall changeStatus"
      id="complete123"
      name="complete123"
      onClick={[Function]}
      title="Mark as New"
      type="button"
    >
      <Image
        alt="Return"
        src="/assets/images/return.png"
      >
        <img
          alt="Return"
          src="/assets/images/return.png"
        />
      </Image>
    </button>
  </Button>
  <Button
    addOnClass="removeToDo"
    buttonId="remove123"
    buttonName="remove123"
    buttonSize="buttonSmall"
    buttonStyle="buttonDanger"
    buttonTitle={
      <Image
        alt="Remove"
        src="/assets/images/trash.png"
      />
    }
    buttonToolTip="Remove ToDo"
    buttonType="button"
    clickEvent={[Function]}
    key="remove123"
  >
    <button
      className="button buttonDanger buttonSmall removeToDo"
      id="remove123"
      name="remove123"
      onClick={[Function]}
      title="Remove ToDo"
      type="button"
    >
      <Image
        alt="Remove"
        src="/assets/images/trash.png"
      >
        <img
          alt="Remove"
          src="/assets/images/trash.png"
        />
      </Image>
    </button>
  </Button>
</ToDoActions>
`;
